pkg/network/tracer/connection/kprobe/tcp_close_consumer.go

```diff
@@ -0,0 +1,126 @@
+// +build linux_bpf
+
+package kprobe
+
+import (
+	"sync/atomic"
+
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	netebpf "github.com/DataDog/datadog-agent/pkg/network/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/network/ebpf/probes"
+	"github.com/DataDog/ebpf/manager"
+)
+
+type tcpCloseConsumer struct {
+	perfHandler  *ddebpf.PerfHandler
+	batchManager *perfBatchManager
+	requests     chan requestPayload
+
+	closedBuffer  *network.ConnectionBuffer
+	maxBufferSize int
+
+	// Telemetry
+	perfReceived int64
+	perfLost     int64
+}
+
+type requestPayload struct {
+	buffer       *network.ConnectionBuffer
+	responseChan chan struct{}
+}
+
+func newTCPCloseConsumer(cfg *config.Config, m *manager.Manager, perfHandler *ddebpf.PerfHandler, filter func(*network.ConnectionStats) bool) (*tcpCloseConsumer, error) {
+	connCloseEventMap, _, err := m.GetMap(string(probes.ConnCloseEventMap))
+	if err != nil {
+		return nil, err
+	}
+	connCloseMap, _, err := m.GetMap(string(probes.ConnCloseBatchMap))
+	if err != nil {
+		return nil, err
+	}
+
+	numCPUs := int(connCloseEventMap.ABI().MaxEntries)
+	batchManager, err := newPerfBatchManager(connCloseMap, numCPUs, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &tcpCloseConsumer{
+		perfHandler:   perfHandler,
+		batchManager:  batchManager,
+		requests:      make(chan requestPayload),
+		closedBuffer:  network.NewConnectionBuffer(1024),
+		maxBufferSize: cfg.MaxClosedConnectionsBuffered,
+	}
+	c.start()
+
+	return c, nil
+}
+
+func (c *tcpCloseConsumer) GetClosedConnections(buffer *network.ConnectionBuffer) int {
+	if buffer == nil {
+		return 0
+	}
+
+	request := requestPayload{
+		buffer:       buffer,
+		responseChan: make(chan struct{}),
+	}
+
+	before := buffer.Len()
+	c.requests <- request
+	<-request.responseChan
+	return buffer.Len() - before
+}
+
+func (c *tcpCloseConsumer) GetStats() map[string]int64 {
+	return map[string]int64{
+		"perf_recv": atomic.SwapInt64(&c.perfReceived, 0),
+		"perf_lost": atomic.SwapInt64(&c.perfLost, 0),
+	}
+}
+
+func (c *tcpCloseConsumer) Stop() {
+	if c == nil {
+		return
+	}
+	c.perfHandler.Stop()
+}
+
+func (c *tcpCloseConsumer) start() {
+	go func() {
+		for {
+			select {
+			case batchData, ok := <-c.perfHandler.DataChannel:
```

Comment:
Good point. I was also wondering if we could write our own perf poller or add an alternative implementation to the ebpf library that uses a buffer pool, because a large chunk of the memory allocations now come from it.